src/data/quaternion.lean

```diff
@@ -0,0 +1,528 @@
+/-
+Copyright (c) 2020 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+
+import tactic.ring_exp ring_theory.algebra algebra.opposites algebra.commute data.equiv.ring
+
+/-!
+# Quaternions
+
+In this file we define quaternions `ℍ[R]` over a commutative ring `R`, and define some
+algebraic structures on `ℍ[R]`.
+
+## Main definitions
+
+* `quaternion_algebra R a b` :
+  [quaternion algebra](https://en.wikipedia.org/wiki/Quaternion_algebra) with coefficients `a`, `b`
+* `quaternion R`, `ℍ[R]` : the space of quaternions, a.k.a. `quaternion_algebra R (-1) (-1)`;
+* `quaternion.norm_sq` : square of the norm of a quaternion;
+* `quaternion.conj` : conjugate of a quaternion;
+
+We also define the following algebraic structures on `ℍ[R]`:
+
+* `ring ℍ[R, a, b]` and `algebra R ℍ[R, a, b]` : for any commutative ring `R`;
+* `ring ℍ[R]` and `algebra R ℍ[R]` : for any commutative ring `R`;
+* `domain ℍ[R]` : for a linear ordered commutative ring `R`;
+* `division_algebra ℍ[R]` : for a linear ordered field `R`.
+
+## Notation
+
+* `ℍ[R]` : quaternions over `R`.
+
+## Implementation notes
+
+We define quaternions over any ring `R`, not just `ℝ` to be able to deal with, e.g., integer
+or rational quaternions without using real numbers. In particular, all definitions in this file
+are computable.
+
+## Tags
+
+quaternion
+-/
+
+/-- Space of quaternions over a type. Implemented as a structure with four fields:
+`re`, `im_i`, `im_j`, and `im_k`. -/
+@[nolint unused_arguments, ext]
+structure quaternion_algebra (R : Type*) (a b : R) :=
+mk {} :: (re : R) (im_i : R) (im_j : R) (im_k : R)
+
+notation `ℍ[` R`,` a`,` b `]` := quaternion_algebra R a b
+
+/-- Space of quaternions over a type. Implemented as a structure with four fields:
+`re`, `im_i`, `im_j`, and `im_k`. -/
+def quaternion (R : Type*) [has_one R] [has_neg R] := quaternion_algebra R (-1) (-1)
+
+notation `ℍ[` R `]` := quaternion R
+
+namespace quaternion_algebra
+
+variables {R : Type*} [comm_ring R] {c₁ c₂ : R} (r x y z : R) (a b c : ℍ[R, c₁, c₂])
+
+instance : has_coe_t R (ℍ[R, c₁, c₂]) := ⟨λ x, ⟨x, 0, 0, 0⟩⟩
+
+mk_simp_attribute quaternion_projs "Projections of operations on quaternions"
```

Comment:
Does this improve things? Is the file slow to compile without? I think we've agreed that `ext; simp; ring_exp` doesn't count as a "non-terminal `simp`".

Reply:
I missed this agreement. I'll try without this.

Reply:
(Okay, maybe I should say I remember @digama0 saying this, and I liked it.)

Reply:
I also agree with Scott's version of me. The important criterion is that `ring` and `ring_exp` doesn't care about the fine details of the goal, for example if it says `a + (b + c)` or `b + (a + c)`, and this is the sort of nondeterminism that usually makes proofs brittle after a `simp`.